### ISSUE TITLE
Potential memory leak in hot reload with JSVM enabled

### ISSUE THREAD

**Issue by lonelycode** [TykTechnologies/tyk#496]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
If I run many hot-reloads (in this case I had removed the timer functionality that introduces limits on how many reloads you can run one after the other) then I can see memory consumption gradually increase with each reload. This behaviour is worse when the JSVM is enabled. 

When the reloads stopped, the RES and VIRT memory figures did not reduce back down

**What is the expected behavior?**
The memory should be reclaimed

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
1. Set ` "reload_wait_time": 1` in tyk.conf
2. Comment out `time.Sleep(10 * time.Second)` in `main.go` in funvtion `doReload()`
3. Compile and run
4. Run a looped script that will call many hot reloads:

```
#!/bin/bash
while [ 1 ]
do
        curl -H "x-tyk-authorization: $SECRET" http://$TYK_GATEWEAY:8181/tyk/reload/group
        sleep .5
done
```

5. Monitor the memory consumption (in this case, without JSVM it required about 2,600 iterations) to from a standing amount (279M or so) to 500M

This isn't so bad as Gateways will not hot reload that often, and when they do, they do not do so very quickly right after each other, however it is concerning that the memory is not being reclaimed by the OS and in the long term could cause issues for the user.

This also affect RPC reloads (MDCB).

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
2.3.2

**Comment by lonelycode** [comment]
On a slightly less vigorous test, with RPC reloads, the memory *was* recouped, but only after a quite long wait. @ilijabojanovic It might be good to get benchmarks for this.

**Comment by mvdan** [comment]
Worth noting that if it takes a while but the memory is eventually freed, it could be simply the GC's tuning.

Will look into the OOM first thing tomorrow.

**Comment by ilijabojanovic** [comment]
@lonelycode ok, also i already planned to start collecting those data since it will be good to have track of results for future releases

**Comment by mvdan** [comment]
@lonelycode does this happen if you do `/tyk/reload`? My assumption is that it does.

In any case, the culprit is very likely here:

```
func ReloadURLStructure() {
        if !reloadScheduled {
                reloadScheduled = true
                go doReload()
                go doCoprocessReload()
                go checkReloadTimeout()
        }
}
```

Assuming coprocess is disabled (`doCoprocessReload does nothing`), it seems like either `doReload` or `checkReloadTimeout` are the culprits. They both do sleeps for some reason, and I don't quite understand why. If the sleeps are long enough, and the reloads triggered fast enough, you could end up with thousands of goroutines before any of the sleeps is over.

Each goroutine takes at minimum 2KB of stack, plus any stack/heap it allocates. So the numbers can escalate to the megabytes very easily. This would explain your OOM.

If we want to make OOMs impossible, we could limit one hot reload at a time (e.g. via a sync.Mutex). I don't see why multiple hot reloads at a time would be useful, anyway. Subsequent requests could either be queued or simply error with "hot reload already in progress".

Note that if there really is a memory leak, this won't solve it. I can see two possibilities for it:

* The sleeps and the GC keep the memory in use for a limited amount of time (is the memory reclaimed in, say, an hour? five hours?)
* There really is a goroutine leak and the memory is in use forever

You didn't specify in the original report, so I'm not sure which it is.

**Comment by mvdan** [comment]
One way to discern which of the two types of leaks it is would be to measure how many reloads it takes to OOM in under a minute (like your mentioned 2,600, rounded up to 3,000 just to be sure) and do them instead over a longer period of time, like every 10s.

This would take much longer, around 8h. Then at the end we would see whether it OOM'd and how much memory is in use.

**Comment by lonelycode** [comment]
>Assuming coprocess is disabled (doCoprocessReload does nothing), it seems like either doReload or checkReloadTimeout are the culprits. They both do sleeps for some reason, and I don't quite understand why. If the sleeps are long enough, and the reloads triggered fast enough, you could end up with thousands of goroutines before any of the sleeps is over.

The sleeps are there to keep a sentinel running so that you can't constantly reload, basically it forces reloads to happen at intervals of x seconds, even if you hit the reload over and over again (with thousands of APIs this becomes a thing, it's irrelevant with small numbers)

We would need to ensure that the hot reload is memory neutral and completely efficient, then we can remove all the timers (think Tyk Cloud - we have thousands of users updating apis all the time, the timers ensure that the stack only reloads in tranches of 120s or so on that system), none of these goroutines execute if `reloadScheduled`.

I would like to clean this up if possible, it's not an ideal situation and more of a kludge.

> One way to discern which of the two types of leaks it is would be to measure how many reloads it takes to OOM in under a minute (like your mentioned 2,600, rounded up to 3,000 just to be sure) and do them instead over a longer period of time, like every 10s.

In my tests it didn't actually OOM kill, I think it started paging.

**Comment by mvdan** [comment]
> I would like to clean this up if possible, it's not an ideal situation and more of a kludge.

It's racy, so I agree that we have to clean it up. Right now, it's unlikely yet possible that two hot reloads start at the same time. We need one of the following:

* A lock around `reloadScheduled`
* A single goroutine handling hot reloads, receiving requests via a channel

I would very much prefer the latter. It's easier to write/maintain and it will be a lot easier to handle throttling of hot reloads. Simply put, if the reloader goroutine is already busy reloading, another reload can't possibly start. And we can limit the amount of reloads per minute/hour too. I assume both of these things would be OK to maintain "you can't constantly reload".

If this sounds fine by you, I'll create a separate issue and assign to myself for 2.4 - in my opinion it's important enough. Hard to tell if it's the cause of this particular issue, though.

**Comment by lonelycode** [comment]
> A single goroutine handling hot reloads, receiving requests via a channel

Go for it, if we can get rid of this dumb timer it will be so much better.

@buger any objections?

**Comment by buger** [comment]
Looks reasonable to me 👍

**Comment by mvdan** [comment]
> Go for it, if we can get rid of this dumb timer it will be so much better.

Awesome :)

> But try to cover this functionality with tests.

Isn't that https://github.com/TykTechnologies/tyk/pull/465? If it's not exactly the same thing, it looks related and should be merged first.

**Comment by mvdan** [comment]
We should try this again on 2.3.x once https://github.com/TykTechnologies/tyk/pull/548 is merged.

If it's still not solved, also worth trying to reproduce on master as there have been other potential fixes.

**Comment by mvdan** [comment]
Fix is merged in both 1.3 and master.

**Comment by buger** [comment]
@mvdan btw, did you checked stuff related to RPC based hot reloads?

**Comment by mvdan** [comment]
I haven't - it seems to use its own loop logic. I'll look around and open an issue if I think it's racey or fragile.

**Comment by mvdan** [comment]
Had a proper look at this. The loop is:

```
func RPCReloadLoop(RPCKey string) {
        for {
                RPCListener.CheckForReload(config.SlaveOptions.RPCKey)
        }
}
```

Which may end up calling:

```
if reload.(bool) {
        // Do the reload!
        log.Warning("[RPC STORE] Received Reload instruction!")
        go signalGroupReload()
        //go ReloadURLStructure()
}
```

This logic seems OK, as it uses no globals and it uses a group reload, which under the hood is a buch of regular reloads which are handled properly.

My only concerns:

* Why is this a busy loop? Don't we want to limit it somewhat to not waste CPU, like the regular reload worker?
* Why is there a commented out `ReloadURLStructure()` in there?

**Comment by mvdan** [comment]
Also funny that `RPCReloadLoop` doesn't actually use its `RPCKey` argument, instead using the global one from the config. Probably harmless since that's the same value it gets passed, but will send a small PR to fix it.

**Comment by mvdan** [comment]
I believe this was fixed when #506 was fixed. @lonelycode @buger have we seen any more reload leaks in 2.3.3/2.3.4?

My comments on https://github.com/TykTechnologies/tyk/issues/496#issuecomment-283093180 are still valid, but I don't think it could lead to a leak. If we'd like to clean that up a little, I can open a separate issue to track those changes.

**Comment by lonelycode** [comment]
We've had reports from users where the JSVM was enabled (enable_jsvm: true) that caused a memory consumption issue - it goes away when you stop the JSVM so it's something to do with the interprester env not being released

**Comment by mvdan** [comment]
Should we open a new issue about that and target for 2.4? Or is it somewhat severe and it should go in a 2.3 bugfix?

**Comment by lonelycode** [comment]
Let's try and get it into a bugfix - 2.4 is a while away and has a big plan which might need whittling down. So either 2.3.5 or 2.3.6, the former would be better obvs...

**Comment by mvdan** [comment]
I assume this JSVM leak is still related to hot reloads, right? As in, it doesn't show at all if there are no reloads?

**Comment by lonelycode** [comment]
Spot on...

**Comment by mvdan** [comment]
Thanks - will get back with findings in the next couple of days.

**Comment by mvdan** [comment]
Running `tyk` at master with the default config, only swapping `enable_jsvm` to `true`, and throwing endless reloads at it (verified it does reload the js VM) does not increase memory usage over time. It does slightly by a couple dozen MBs, but the GC catches up in a few seconds and it goes back down. Left it running for a couple of minutes to no noticeable increase in memory usage.

Am I doing anything wrong? Perhaps I should actually add a js middleware or run requests through it?

Another thought is that this leak could be caused because the old VM keeps running. We don't keep any pointers to the old VM as far as I can tell, but it could be that it doesn't get garbage collected because it's still running in an http request goroutine.

For example, say I write a js middleware that just sleeps forever. I assume that a request going through it would either never reply, or time out and give an error response. Either way, the js VM will keep on running. We do nothing to stop it, right now.

Otto provides a way to kill the VM: https://godoc.org/github.com/robertkrimen/otto#hdr-Halting_Problem I haven't been able to reproduce the leak so I'm not sure if this could be a fix, but it could be.

I can submit a PR to kill the previous Otto VM when starting up a new one as part of a hot reload. The only question is: is that safe? If the old router and old Otto VM are still processing a request, this could lead to killing it. Do we want to wait until they're done doing all their work or until a timeout?

**Comment by mvdan** [comment]
@ilijabojanovic if you've played with js middleware before, perhaps you can reproduce this memory leak.

@matiasinsaurralde any ideas, since you wrote most of this code?

**Comment by matiasinsaurralde** [comment]
@mvdan I didn't write the JS plugin support 😢 
I've seen some memory issues in the Otto repo, perhaps you've seen them already.

**Comment by mvdan** [comment]
My bad - thought it was part of coprocess :)

You're probably referencing to this, right? https://github.com/robertkrimen/otto/issues/71

I hadn't noticed the `Copy()`, thanks for the input. So perhaps I do need to throw requests at a js middleware to trigger the leak. If the issue seems like an upstream Otto bug, I'll update that issue.

**Comment by mvdan** [comment]
By the way @matiasinsaurralde, what you said in that issue - `memory keeps growing on each request` - you didn't do any reloads to reproduce the leak, right?

**Comment by matiasinsaurralde** [comment]
@mvdan JS support through Otto came first, I wrote the CP based on the JS middleware structure.

Yes, it's that issue! I haven't tried to reproduce it. I've found it on the Otto repo a long time ago but there was no Tyk related bug report back then.

**Comment by mvdan** [comment]
Okay, given that I couldn't reproduce anything with just reloads, my chips are on the cause being either `Copy()` or that we never kill the copied VMs if they never finish (both related to requests, not reloads).

Thanks for your help!

**Comment by mvdan** [comment]
I've done another version of the test I did above - which was throwing endless reloads on a Tyk with the JSVM on - but this time also with an endpoint that has a js pre middleware that just sets a header.

It has been running for over 20m now, counting around a thousand requests and a few hundred reloads. Memory usage is still stable at 37-38MB, with no leak to be seen.

So I think the original post by Martin about reloads with near-zero timeouts was simply a race that was leaving goroutines hanging and was fixed in #506.

As to the JSVM leaks that Martin mentioned later in the thread and Matias said he encountered too, I haven't been able to reproduce them either. Perhaps they have been fixed somehow too, I'm not sure. In any case, there is still a potential leak when it comes to JSVM copies not stopping by themselves which we must kill, as I outlined above.

I'll prepare a pull request killing stalled JSVM copies which we can include in 2.3.5. Unless someone can reproduce a leak on master or 2.3.4 with reloads and/or the JSVM, I think we should close or postpone this (once the PR mentioned is merged) and revisit if one of us or a customer encounters a leak again.
